bin/elasticurl/main.cpp

```diff
@@ -0,0 +1,540 @@
+    fprintf(stderr, "      --connect-timeout INT: time in milliseconds to wait for a connection.\n");
+    fprintf(stderr, "  -H, --header LINE: line to send as a header in format [header-key]: [header-value]\n");
+    fprintf(stderr, "  -d, --data STRING: Data to POST or PUT\n");
+    fprintf(stderr, "      --data-file FILE: File to read from file and POST or PUT\n");
+    fprintf(stderr, "  -M, --method STRING: Http Method verb to use for the request\n");
+    fprintf(stderr, "  -G, --get: uses GET for the verb.\n");
+    fprintf(stderr, "  -P, --post: uses POST for the verb.\n");
+    fprintf(stderr, "  -I, --head: uses HEAD for the verb.\n");
+    fprintf(stderr, "  -i, --include: includes headers in output.\n");
+    fprintf(stderr, "  -k, --insecure: turns off SSL/TLS validation.\n");
+    fprintf(stderr, "  -o, --output FILE: dumps content-body to FILE instead of stdout.\n");
+    fprintf(stderr, "  -t, --trace FILE: dumps logs to FILE instead of stderr.\n");
+    fprintf(stderr, "  -v, --verbose: ERROR|INFO|DEBUG|TRACE: log level to configure. Default is none.\n");
+    fprintf(stderr, "      --version: print the version of elasticurl.\n");
+    fprintf(stderr, "      --http2: HTTP/2 connection required");
+    fprintf(stderr, "      --http1_1: HTTP/1.1 connection required");
+    fprintf(stderr, "  -h, --help\n");
+    fprintf(stderr, "            Display this message and quit.\n");
+    exit(exit_code);
+}
+
+static struct aws_cli_option s_long_options[] = {
+    {"cacert", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'a'},
+    {"capath", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'b'},
+    {"cert", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'c'},
+    {"key", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'e'},
+    {"connect-timeout", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'f'},
+    {"header", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'H'},
+    {"data", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'd'},
+    {"data-file", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'g'},
+    {"method", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'M'},
+    {"get", AWS_CLI_OPTIONS_NO_ARGUMENT, NULL, 'G'},
+    {"post", AWS_CLI_OPTIONS_NO_ARGUMENT, NULL, 'P'},
+    {"head", AWS_CLI_OPTIONS_NO_ARGUMENT, NULL, 'I'},
+    {"include", AWS_CLI_OPTIONS_NO_ARGUMENT, NULL, 'i'},
+    {"insecure", AWS_CLI_OPTIONS_NO_ARGUMENT, NULL, 'k'},
+    {"output", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'o'},
+    {"trace", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 't'},
+    {"verbose", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'v'},
+    {"version", AWS_CLI_OPTIONS_NO_ARGUMENT, NULL, 'V'},
+    {"http2", AWS_CLI_OPTIONS_NO_ARGUMENT, NULL, 'w'},
+    {"http1_1", AWS_CLI_OPTIONS_NO_ARGUMENT, NULL, 'W'},
+    {"help", AWS_CLI_OPTIONS_NO_ARGUMENT, NULL, 'h'},
+    /* Per getopt(3) the last element of the array has to be filled with all zeros */
+    {NULL, AWS_CLI_OPTIONS_NO_ARGUMENT, NULL, 0},
+};
+
+static void s_parse_options(int argc, char **argv, struct elasticurl_ctx *ctx)
```

Comment:
You **basically never** pass arguments by raw pointer in c++. Usually you pass by reference, ex: `ElasticurlCtx &ctx`. A reference and a pointer are the same thing as far as the compiled code is concerned, but a reference can never be NULL.
The only times you'd still use raw pointers are:
- If the argument is OPTIONAL
- `**argv` because it's as old as the dinosaurs
- `const char *` for non-owning strings (c++17 introduced std::string_view, but we're limiting ourselves to c++14)

Oh yeah, and you don't need to write `struct` except where the struct is initially declared.